### DISCUSSION TITLE
JOIN against a table with inheritance results in a slow query

### DISCUSSION THREAD

**Discussion by mat105** [sqlalchemy/sqlalchemy#10876] [Usage Questions]
Hi
I have a complex query in postgres 15, where i have a table deliveries and a table addresses.
an address can be a warehouse, an agency or a pudo (pickup point).

I make the join this way:

```py
origin = with_polymorphic(
    AddressModel, [WarehouseModel, PudoModel, AgencyModel], aliased=True, flat=True
)
destination = with_polymorphic(
    AddressModel, [WarehouseModel, PudoModel, AgencyModel], aliased=True, flat=True
)

query = (
    select(DeliveryModel)
    .join(origin, DeliveryModel.origin_address)
    .join(destination, DeliveryModel.destination_address)
)

query = query.options(
    contains_eager(DeliveryModel.origin_address.of_type(origin))
    contains_eager(DeliveryModel.destination_address.of_type(destination))
)
```
However the previous query when executed does the inheritance join inside parenthesis, and does the JOIN against deliveries outside. For some reason doing this, the inheritance join are using a sequential SCAN instead of an INDEX scan, resulting on a really slow query.

```sql
FROM deliveries
JOIN (addresses AS addresses_1
      LEFT OUTER JOIN warehouses AS warehouses_1 ON addresses_1.id = warehouses_1.id
      LEFT OUTER JOIN pudos AS pudos_1 ON addresses_1.id = pudos_1.id
      LEFT OUTER JOIN agencies AS agencies_1 ON addresses_1.id = agencies_1.id) ON deliveries.origin_address_id = addresses_1.id
JOIN (addresses AS addresses_2
      LEFT OUTER JOIN warehouses AS warehouses_2 ON addresses_2.id = warehouses_2.id
      LEFT OUTER JOIN pudos AS pudos_2 ON addresses_2.id = pudos_2.id
      LEFT OUTER JOIN agencies AS agencies_2 ON addresses_2.id = agencies_2.id) ON deliveries.destination_address_id = addresses_2.id

```
If i replace the parenthesis, and put the ON in the first join against addresses the query goes much faster.
`JOIN addresses AS addresses_1 ON deliveries.origin_address_id = addresses_1.id`

Is there some way for the ON condition to be done on addresses instead of outside?

**Comment by zzzeek** [comment]
if you are building with_polymorphic() constructs, those necessarily are going to group their JOIN expressions, these are LEFT OUTER JOINs so those are not commutative against an inner JOIN.

there's no actual need to use with_polymorphic to actually run the SELECT statement if all the criteria you are setting up is in terms of the "addresses" table so I'd say dont use with_polymoprhic here.    also the join() and contains_eager() illustrated don't seem to match up.
